### DISCUSSION TITLE
How to determine the cost of deploying bundles?

### DISCUSSION THREAD

**Discussion by 0xpatrickdev** [Agoric/agoric-sdk#8285] [Q&A] [answered]
I am trying to figure out the cost for deploying bundles, via something like `agd tx swingset install-bundle` or `agoric publish`. I was pointed to the [DefaultBeansPerStorageByte parameter](https://github.com/Agoric/agoric-sdk/blob/ac9f916a844f06db1c43e9cc69a948d9c3528a4e/golang/cosmos/x/swingset/types/default-params.go#L51), which currently seems to be set to 0.002. 

Using the [addCollateral bundles](https://github.com/0xpatrickdev/agoric-vault-collateral-proposal/tree/main/bundles) from inter protocol as a reference point, I have come up with these numbers:

| Hash            | Size           | Cost per Byte | Cost   |
|-----------------|----------------|---------------|-----------|
| `b1-1c8e93...`  | 948,173 bytes  | 0.002        | 1,896.35 |
| `b1-8fb229...`  | 1,456,129 bytes| 0.002        | 2,912.26 |
| `b1-c185bf...`  | 1,465,213 bytes| 0.002        | 2,930.43 |
| `b1-e4ba9c...`  | 1,114,541 bytes| 0.002        | 2,229.08 |
| **Total**       | **4,984,056 bytes** |   | 9,968.11 |

 - These numbers seem a bit high, so I would I appreciate if someone could double check the calculation. I am assuming that the total is ~10,000 IST.
 - Also, is there any way to query the value of `DefaultBeansPerStorageByte` from agd?

**Comment by dckc** [comment, marked as answer]
## Contract Bundle Storage Currently 0.002 IST/byte

> I was pointed to the [DefaultBeansPerStorageByte parameter](https://github.com/Agoric/agoric-sdk/blob/ac9f916a844f06db1c43e9cc69a948d9c3528a4e/golang/cosmos/x/swingset/types/default-params.go#L51), which currently seems to be set to 0.002.

Yes, it's effectively **0.002 IST/byte** currently. The value of that `storageByte` param is actually 2000000000 in units of "beans", and the ratio of beans to IST is `feeUnit` = 1000000000000.

## Storage Price Subject To Governance

> These numbers seem a bit high ...

I say "currently" because it can be changed by the BLD stakers using a [Parameter Change governance proposal](https://hub.cosmos.network/main/governance/proposal-types/param-change.html).

The value should be sufficient to recover the the cost of storing bundles on the chain forever / indefinitely.

For comparison, storage on Ethereum currently costs about 10x as much:

  | gas | bytes | USD | gas price (gwei) | ETH price
-- | -- | -- | -- | -- | --
eth | 625 | 1 | 0.0223810125 | 22 | 1,627.71

 - 20k gas / 256 bit word (32 bytes)
 - gas price, eth price from https://etherscan.io/gastracker

### Storage fees go to the reserve account

The `vbank/reserve` module account is `agoric1ae0lmtzlgrcnla9xjkpaarq5d5dfez63h3nucl`.

[on agoricdev-20](https://devnet.explorer.agoric.net/agoric/account/agoric1ae0lmtzlgrcnla9xjkpaarq5d5dfez63h3nucl) it has 16,407.6 IST from recent deployment of bundles.

## Querying Storage Price

> is there any way to query the value of `DefaultBeansPerStorageByte` from agd?

yes: `agd query swingset params`.

If you don't have a local node running, for mainnet, you could use: `agd --node=https://main.rpc.agoric.net:443 query swingset params`

to get just the per-byte cost: `agd --node=https://main.rpc.agoric.net:443 query swingset params -o json | jq '.beans_per_unit | .[] | select(.key=="storageByte")'`
```json
{
  "key": "storageByte",
  "beans": "2000000000"
}
```

The `feeUnit` lets us translate to IST:
`gd --node=https://main.rpc.agoric.net:443 query swingset params -o json | jq '.beans_per_unit | .[] | select(.key=="storageByte" or .key=="feeUnit")'`

```json
{
  "key": "feeUnit",
  "beans": "1000000000000"
}
{
  "key": "storageByte",
  "beans": "2000000000"
}
```

**Comment by dckc** [comment]
## How to reduce bundle size?

In due course, you should only need to pay for new modules:

 - https://github.com/Agoric/agoric-sdk/issues/2391

in the shorter term, there's some possible clean-up we are looking into:

 - https://github.com/Agoric/agoric-sdk/issues/6512

**Comment by dckc** [comment]
TIL:
> On November 22, 2016 the Spurious Dragon hard-fork introduced EIP-170 which added a smart contract size limit of 24.576 kb.
-- [Downsizing contracts to fight the contract size limit \| ethereum\.org](https://ethereum.org/en/developers/tutorials/downsizing-contracts-to-fight-the-contract-size-limit/)

[using `for`](https://docs.soliditylang.org/en/v0.6.10/contracts.html#using-for) also seems interesting

**Comment by dckc** [comment]
Also on the topic of storage cost...

 - 2018-11: [EIP\-1682: Storage Rent](https://eips.ethereum.org/EIPS/eip-1682) (withdrawn)
 - 2021-03: [spacebank: vat storage rental economics · Issue \#2631 · Agoric/agoric\-sdk](https://github.com/Agoric/agoric-sdk/issues/2631)
   - including some notes about Solana storage rental

**Comment by dckc** [comment]
## 99% reduction under discussion

Speaking of Subject to Governance:

 - [Proposal to Reduce the Cost of Installing a Bundle on Agoric Blockchain \- Governance \- Agoric Community Forum](https://community.agoric.com/t/proposal-to-reduce-the-cost-of-installing-a-bundle-on-agoric-blockchain/523)

## UI for on-chain proposal

An on-chain proposal involves just the right `agd` command with just the right .json file. The first time I/we did it, we goofed and had to fix it in a follow-up proposal:

 - [Proposal \#7 Fix formatting of governance parameters](https://agoric.explorers.guru/proposal/7)
   - [fix-gov-params.json](https://gist.github.com/dckc/ea859aeea1f8864da2abbb1c6a8ce51d#file-fix-gov-params-json)
   - [Agoric – Staking Rewards and Genesis Parameter Updates](https://commonwealth.im/agoric/discussion/5797-staking-rewards-and-genesis-parameter-updates)

So this time, it would be nice to have a friendly UI for it. To that end, I opened an issue on the nifty new gov-proposal-hub:

 - https://github.com/0xpatrickdev/agoric-gov-proposal-hub/issues/10

And this weekend I mostly got it working:

 - https://github.com/0xpatrickdev/agoric-gov-proposal-hub/pull/18

![image](https://github.com/Agoric/agoric-sdk/assets/150986/33570e15-fe72-40bb-bb8d-b8d06303bccc)

## Reference Doc for Swingset Params

It's still a little opaque. It seems like there should be some reference documentation for things like **storageByte**. I can't find any. All I find is the code of the [chargeAdmission](https://github.com/Agoric/agoric-sdk/blob/0e22bfac38fa6073c8f8c9a5de62ff91c447e44c/golang/cosmos/x/swingset/types/msgs.go#L39) function along with the PR and issue whence it came:

 - https://github.com/Agoric/agoric-sdk/issues/6858
 - https://github.com/Agoric/agoric-sdk/pull/7290

The [swingset Params message](https://github.com/Agoric/agoric-sdk/blob/master/golang/cosmos/proto/agoric/swingset/swingset.proto#L36) doesn't specify `storageByte`. It just has a `beans_per_unit` array in which `storageByte` is one of the items.
